### ISSUE TITLE
Support layout renderer in log level filters (e.g. minLevel, maxLevel)

### ISSUE THREAD

**Issue by Matthew-Davey** [NLog/NLog#1502]
- Feature request

**NLog version**:  4.3.3)

**Platform**: .Net 4.0 , Mono4

I'd really love to be able to use layout renderers for my log level. Specifically my use case is - I need to be able to specify a default log level, but have this be overridable with an environment variable.

I tried the following but unfortunately I did not see any log output at all :(

``` xml
<rules>
  <logger name="*" minLevel="${whenEmpty:whenEmpty=TRACE:inner=${environment:variable=ST_LOGLEVEL}}" appendTo="console, syslog" />
</rules>
```

and this:

``` xml
<rules>
      <logger name="*" minLevel="${environment:variable=ST_LOGLEVEL:whenEmpty=TRACE}" appendTo="console, syslog" />
    </rules>
```

and finally this:

``` xml
<rules>
      <logger name="*" minLevel="${literal:text=TRACE}" appendTo="console, syslog" />
    </rules>
```

Which led me to believe that layout renderers are not available at all for log level.

**Comment by 304NotModified** [comment]
That is indeed currently not supported, but would be powerful. I would expect this will cost some performance, but probable we can make it still fast in the "old" way.

**Comment by 304NotModified** [comment]
You can work around it by using filters. https://github.com/NLog/NLog/wiki/When-filter

Bit more verbose but powerful

**Comment by snakefoot** [comment]
Not an easy nut to crack, as the logging-rules has to be super-efficient, and calling layout-renderers and allocating strings is not super-efficient.

But while waiting, then one might use this:

https://stackoverflow.com/questions/47475698/how-to-control-nlog-loglevel-using-environment-variables/47517826#47517826

**Comment by snakefoot** [comment]
When #2709 has been implemented, then one should be able to just modify the nlog.config (with autoreload=true), and the logging rules will adapt.

Not exactly the implementation the original issue poster asked for, but this will work fast and will provide dynamic change of loglevel without application restart.

**Comment by 304NotModified** [comment]
> Not an easy nut to crack, as the logging-rules has to be super-efficient, and calling layout-renderers and allocating strings is not super-efficient.

Therefor it in #2709 a only interpreted when reading the config :)

**Comment by snakefoot** [comment]
A workaround could be to lookup a LoggingRule using its RuleName. And the modify the LoggingRule Level-filter directly. See also #2891

But else having autoReload="true" together with #2709 will also update the Level-filter by reloading the entire LoggingConfiguration.

**Comment by 304NotModified** [comment]
fixed by #3184 :)
